core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java

```diff
@@ -994,6 +1006,139 @@ public void date_format() {
+    }
+
+    assertEquals(nullValue(), eval(maketime(DSL.literal(-1.), DSL.literal(42.), DSL.literal(42.))),
+        "Negative hour doesn't produce NULL");
+    assertEquals(nullValue(), eval(maketime(DSL.literal(42.), DSL.literal(-1.), DSL.literal(42.))),
+        "Negative minute doesn't produce NULL");
+    assertEquals(nullValue(), eval(maketime(DSL.literal(42.), DSL.literal(42.), DSL.literal(-1.))),
+        "Negative second doesn't produce NULL");
+
+    // The function has 3 arguments, testing NULL and MISSING in all position(s)
+    var combinations = permuteArgumentsWith(3, nullRef, DSL.literal(42.));
+    for (var args : combinations) {
+      assertEquals(nullValue(), eval(maketime(args[0], args[1], args[2])));
+    }
+    combinations = permuteArgumentsWith(3, missingRef, DSL.literal(42.));
+    for (var args : combinations) {
+      assertEquals(missingValue(), eval(maketime(args[0], args[1], args[2])));
+    }
+  }
+
+  private FunctionExpression makedate(Expression year, Expression dayOfYear) {
+    var repo = new ExpressionConfig().functionRepository();
+    var func = repo.resolve(new FunctionSignature(new FunctionName("makedate"),
+        List.of(DOUBLE, DOUBLE)));
+    return (FunctionExpression)func.apply(List.of(year, dayOfYear));
+  }
+
+  private LocalDate makedate(Double year, Double dayOfYear) {
+    return makedate(DSL.literal(year), DSL.literal(dayOfYear)).valueOf(null).dateValue();
+  }
+
+  @Test
+  public void makedate() {
+    var r = new Random();
+    for (var ignored : IntStream.range(0, 1025).toArray()) {
+      // Avoid zero values
+      var year = r.nextDouble() * 5000 + 1;
+      var dayOfYear = r.nextDouble() * 1000 + 500;
+
+      LocalDate actual = makedate(year, dayOfYear);
+
+      // Using another algorithm to get reference value
+      // We should go to the next year until remaining @dayOfYear is bigger than 365/366
+      var yearL = (int)Math.round(year);
+      var dayL = (int)Math.round(dayOfYear);
+      while (true) {
+        int daysInYear = Year.isLeap(yearL) ? 366 : 365;
+        if (dayL > daysInYear) {
+          dayL -= daysInYear;
+          yearL++;
+        } else {
+          break;
+        }
+      }
+      LocalDate expected = LocalDate.ofYearDay(yearL, dayL);
+
+      assertEquals(expected, actual,
+          String.format("year = %f, dayOfYear = %f", year, dayOfYear));
+    }
+
+    assertEquals(LocalDate.ofYearDay(2000, 42), makedate(0., 42.),
+        "0 year is not interpreted as 2000 as in MySQL");
+    assertEquals(nullValue(), eval(makedate(DSL.literal(-1.), DSL.literal(42.))),
+        "Negative year doesn't produce NULL");
+    assertEquals(nullValue(), eval(makedate(DSL.literal(42.), DSL.literal(-1.))),
+        "Negative dayOfYear doesn't produce NULL");
+    assertEquals(nullValue(), eval(makedate(DSL.literal(42.), DSL.literal(0.))),
+        "Zero dayOfYear doesn't produce NULL");
```

Comment:
what about tests for invalid hour/minute/second or year/month/day? 
(is there an invalid year, or is year 0 and 9999 valid?)

Reply:
MySQL is limited on output value: from 0001-01-01 to 9999-12-31. Java has no limits.